### DISCUSSION TITLE
Design patterns for industry quality ASR

### DISCUSSION THREAD

**Discussion by KacperKubara** [speechbrain/speechbrain#1822] [General]
Hi! 

For the past couple of years, I've had an opportunity to work with speech recognition in a research and more applied environment. Throughout that time, I've seen a few different approaches how to deploy those models on the cloud and making them usable. 

I was wondering what architectural design patterns did you encounter for ASR? Also, how do you ensure low latency for near real-time transcriptions? How would you design it to be deployable on the cloud?

_Some of the challenges that I observed with ASR models:_
* If you use Docker containers, models should be downloaded while the container is created (so the Docker image is kept small).  This makes the startup of the containers quite slow which should be kept in mind if you need to spin more containers due to a higher user load.

* Near real-time transcription is HARD with ASR models. You might need to choose smaller models with higher WER that are faster and go for C++-based libraries such as Kaldi to get a better speed.

_Designs that I've encountered so far:_
* Out-of-the-box solutions: software built around Speech Services API such as Azure Cognitive Services. You can make it a bit more elaborate and allow you to finetune the cloud-native models to a custom domain as well. This is EXPENSIVE and not very accurate, more or less ~10 times more expensive than hosting your model in a container or VM.
* VMs: set up some virtual machines on the cloud and host the models there with exposed API, through Flask, for example. This is relatively to start with on the cloud but not a particularly cost-efficient and scalable solution.
* Microservices-based architecture: In my previous startup we had several models that were using transcription as an input (e.g. summarization). We containerized each model and made them communicate with each other by a queue. That was quite a nice solution as it isolated the components and made the more scalable. However, it was much more difficult to debug it as we didn't have good monitoring in place. I attach a diagram below for this architecture: 

![image](https://user-images.githubusercontent.com/33067446/215709174-1049a87f-1339-4089-9316-916f63b4c90a.png)

**Comment by asumagic** [comment]
> I was wondering what architectural design patterns did you encounter for ASR? Also, how do you ensure low latency for near real-time transcriptions?

I have already implemented a relatively monolithic solution (`aiorpc`+pykaldi), but that is probably not too relevant to larger scale cloud deployment.

The biggest gain you can have for latency is probably to use models that support streaming, which is generally not the case for larger transformer-based models, but that's often a tradeoff with accuracy as you said.

When using such models, as you need to feed an entire input sequence to the system, the latency/processing overhead comes more from having to either continuously feed large overlapping sequences of the audio history, or to having to identify complete utterances using VAD.

(A subset of) RNN-T models are streaming-friendly, relatively lightweight and appears to be suitable for realtime transcription (e.g. Google's on-device ASR uses RNN-T models nowadays AFAIK). When I looked there was quite a bit of promising research and possible improvements, but I don't know how much of these have been implemented and real world tested in FOSS toolkits. Generally speaking, SpeechBrain does not have out-of-the-box/convenient support for online streaming though.
